llvm_passes/HipPrintf.cpp

```diff
@@ -18,158 +53,363 @@
+    Arg = ASCast->getPointerOperand();
+  } else if (auto CE = dyn_cast<ConstantExpr>(Arg)) {
+    if (CE->getOpcode() == llvm::Instruction::AddrSpaceCast) {
+      Arg = CE->getOperand(0);
+    } else {
+      assert("Cannot handle the printf format string!" && false);
+    }
   }
+  return cast<GlobalVariable>(Arg);
+}
+
+// Extracts the format string in the printf argument passed as FmtStrArg,
+// and (possibly) splits it to a number of format strings if there are %s
+// format args. In such a case the %s format string is returned as a separate
+// "%s" format string. Also counts the total number of format specifiers in
+// NumberOfFormatSpecs to returns it as a final return value.
+static std::vector<std::string>
+getFormatStringPieces(Value *FmtStrArg, unsigned &NumberOfFormatSpecs) {
+
+  Type *Int8Ty = IntegerType::get(FmtStrArg->getContext(), 8);
+
+  ConstantExpr *CE = cast<ConstantExpr>(FmtStrArg);
+
+  Value *FmtStrOpr = findGlobalStr(CE->getOperand(0));
 
-  StringRef getPassName() const override {
-    return "Convert printf calls to OpenCL compatible ones.";
+  GlobalVariable *OrigFmtStr = findGlobalStr(CE->getOperand(0));
+
+  std::vector<std::string> FmtStrPieces;
+  ConstantDataSequential *FmtStrData =
+      dyn_cast<ConstantDataSequential>(OrigFmtStr->getInitializer());
+
+  if (FmtStrData == nullptr) {
+    assert(OrigFmtStr->getInitializer()->isZeroValue());
+    FmtStrPieces.push_back("");
+    NumberOfFormatSpecs = 0;
+    return FmtStrPieces;
   }
-};
 
-char HipPrintfToOpenCLPrintfLegacyPass::ID = 0;
-static RegisterPass<HipPrintfToOpenCLPrintfLegacyPass>
-    X("hip-printf", "Convert printf calls to OpenCL compatible ones.");
+  auto FmtString = FmtStrData->getAsString();
+
+  // Counting the format specs is a more accurate approximation for the return
+  // value than counting the passed arguments due to the dynamic width
+  // specifier (*) which is given as an argument and should not be counted
+  // in to the processed arguments (I believe).
+  NumberOfFormatSpecs = NumFormatSpecs(FmtString);
+
+  // FmtStrData has one byte larger size since it includes the trailing
+  // byte, so if we convert it directly to a std::string, it gets a corrupted
+  // string due to the wrong size passed to it (occurs at least with LLVM 13)?
+  // TO CHECK again.
+  std::string FmtStr = std::string(FmtStrData->getAsCString());
+
+  // TODO: %s can also have * specifier. We have to treat it as a special
+  // case, adding strlen() - width for padding.
+  size_t Pos = 0;
+  while ((Pos = FmtStr.find("%s")) != std::string::npos &&
+         // Without this we'd handle %%s wrongly.
+         (Pos > 1 && FmtStr[Pos - 1] != '%')) {
+    std::string TokenBefore = FmtStr.substr(0, Pos);
+    if (TokenBefore != "") {
+      FmtStrPieces.push_back(TokenBefore);
+    }
+    assert(FmtStr.size() > Pos + 2 - 1);
+    FmtStr.erase(0, Pos + 2);
+    FmtStrPieces.push_back("%s");
+  }
+  if (FmtStr != "") {
+    FmtStrPieces.push_back(FmtStr);
+  }
```

Comment:
```suggestion
  if (FmtStr != "")
    FmtStrPieces.push_back(FmtStr);
```
(https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements)

Reply:
There are more cases like this.